File: smoke_tests/java/RegexDos.java
// Regular Expression DoS (ReDoS) vulnerabilities in Java

import java.util.regex.Pattern;
import java.util.regex.Matcher;
import javax.servlet.http.*;

public class RegexDos extends HttpServlet {

    // Test 1: Nested quantifiers
    public boolean validateNested(String input) {
        // VULNERABLE: (a+)+ causes exponential backtracking
        Pattern pattern = Pattern.compile("(a+)+b");
        return pattern.matcher(input).matches();
    }

    // Test 2: Overlapping alternation
    public boolean checkOverlap(String input) {
        // VULNERABLE: Overlapping alternatives
        Pattern pattern = Pattern.compile("(a|a)+b");
        return pattern.matcher(input).matches();
    }

    // Test 3: Email validation ReDoS
    public boolean validateEmail(String email) {
        // VULNERABLE: Classic email ReDoS pattern
        String regex = "^([a-zA-Z0-9]+)+@([a-zA-Z0-9]+)+\\.([a-zA-Z]+)+$";
        return email.matches(regex);
    }

    // Test 4: URL validation ReDoS
    public boolean validateUrl(String url) {
        // VULNERABLE: Nested groups in URL pattern
        Pattern pattern = Pattern.compile("^(https?://)?([a-zA-Z0-9.-]+)+(/.*)* $");
        return pattern.matcher(url).matches();
    }

    // Test 5: User-supplied regex
    public boolean customMatch(HttpServletRequest request) {
        String input = request.getParameter("input");
        String patternStr = request.getParameter("pattern");
        // VULNERABLE: User supplies regex pattern
        Pattern pattern = Pattern.compile(patternStr);
        return pattern.matcher(input).matches();
    }

    // Test 6: replaceAll with vulnerable pattern
    public String replacePattern(String input) {
        // VULNERABLE: Vulnerable pattern in replaceAll
        return input.replaceAll("(a+)+", "X");
    }

    // Test 7: split with ReDoS
    public String[] splitInput(String input) {
        // VULNERABLE: Split can trigger ReDoS
        return input.split("(\\s+)+");
    }

    // Test 8: HTML tag matching ReDoS
    public int matchHtml(String html) {
        // VULNERABLE: Complex HTML pattern
        Pattern pattern = Pattern.compile("<([a-z]+)([^>]*)*>");
        Matcher matcher = pattern.matcher(html);
        int count = 0;
        while (matcher.find()) {
            count++;
        }
        return count;
    }

    // Test 9: Multiline ReDoS
    public boolean multilineMatch(String content) {
        // VULNERABLE: Multiline with backtracking
        Pattern pattern = Pattern.compile("^(.+)+$", Pattern.MULTILINE);
        return pattern.matcher(content).matches();
    }

    // Test 10: Case insensitive ReDoS
    public boolean caseInsensitive(String input) {
        // VULNERABLE: Case insensitive can make it worse
        Pattern pattern = Pattern.compile("([a-z]+)+$", Pattern.CASE_INSENSITIVE);
        return pattern.matcher(input).matches();
    }
}

File: smoke_tests/ruby/regex_dos.rb
# Regular Expression DoS (ReDoS) vulnerabilities in Ruby
class RegexController < ApplicationController
  # Test 1: Nested quantifiers
  def validate_nested
    input = params[:input]
    # VULNERABLE: (a+)+ causes exponential backtracking
    pattern = /(a+)+b/
    render json: { match: pattern.match?(input) }
  end

  # Test 2: Overlapping alternation
  def check_overlap
    input = params[:input]
    # VULNERABLE: Overlapping alternatives
    pattern = /(a|a)+b/
    render json: { match: pattern.match?(input) }
  end

  # Test 3: Email validation ReDoS
  def validate_email
    email = params[:email]
    # VULNERABLE: Classic email ReDoS pattern
    pattern = /^([a-zA-Z0-9]+)+@([a-zA-Z0-9]+)+\.([a-zA-Z]+)+$/
    render json: { valid: pattern.match?(email) }
  end

  # Test 4: URL validation ReDoS
  def validate_url
    url = params[:url]
    # VULNERABLE: Nested groups in URL pattern
    pattern = /^(https?:\/\/)?([a-zA-Z0-9.-]+)+(\/.*)*$/
    render json: { valid: pattern.match?(url) }
  end

  # Test 5: User-supplied regex
  def custom_match
    input = params[:input]
    pattern_str = params[:pattern]
    # VULNERABLE: User supplies regex pattern
    pattern = Regexp.new(pattern_str)
    render json: { match: pattern.match?(input) }
  end

  # Test 6: gsub with vulnerable pattern
  def replace_pattern
    input = params[:input]
    # VULNERABLE: Vulnerable pattern in gsub
    result = input.gsub(/(a+)+/, 'X')
    render plain: result
  end

  # Test 7: scan with backtracking
  def scan_input
    input = params[:input]
    # VULNERABLE: Scan with exponential pattern
    matches = input.scan(/(.+)+x/)
    render json: { count: matches.length }
  end

  # Test 8: split with ReDoS
  def split_input
    input = params[:input]
    # VULNERABLE: Split can trigger ReDoS
    parts = input.split(/(\s+)+/)
    render json: { parts: parts }
  end

  # Test 9: HTML tag matching ReDoS
  def match_html
    html = params[:html]
    # VULNERABLE: Complex HTML pattern
    pattern = /<([a-z]+)([^>]*)*>/
    matches = html.scan(pattern)
    render json: { count: matches.length }
  end

  # Test 10: Multiline ReDoS
  def multiline_match
    content = params[:content]
    # VULNERABLE: Multiline with backtracking
    pattern = /^(.+)+$/m
    render json: { match: pattern.match?(content) }
  end

  # Test 11: Regex from database
  def dynamic_regex
    config = PatternConfig.find(params[:id])
    input = params[:input]
    # VULNERABLE: Pattern from database (could be user-supplied)
    pattern = Regexp.new(config.pattern)
    render json: { match: pattern.match?(input) }
  end

  # Test 12: Case insensitive ReDoS
  def case_insensitive
    input = params[:input]
    # VULNERABLE: Case insensitive can make it worse
    pattern = /([a-z]+)+$/i
    render json: { match: pattern.match?(input) }
  end
end
